### DISCUSSION TITLE
subscribeWithSelector doesn't fire or notify only on state changes that's passed to subscribe

### DISCUSSION THREAD

**Discussion by guenyoo** [pmndrs/zustand#1378] [Q&A]
Hey folks,

I like Zustand a lot and would actually need to use it in between React and VanillaJS, however I'm running into problems. Am I using it incorrectly or is something not working correctly?

I'm referencing to this part of the documentation:

[Reading/writing state and reacting to changes outside of components](https://docs.pmnd.rs/zustand/recipes/recipes#reading/writing-state-and-reacting-to-changes-outside-of-components)

[I've also created a codesandbox to illustrate the problem:](https://codesandbox.io/s/npm-playground-forked-kkpxbc?file=/src/index.js)

It's very basic:
```javascript
import create from "zustand";
import { subscribeWithSelector } from "zustand/middleware";
const useStoreWithSelector = create(
  subscribeWithSelector(() => ({ paw: true, snout: true, fur: true }))
);

useStoreWithSelector.subscribe(console.log, (state) => state.paw);

useStoreWithSelector.setState({ paw: false });
useStoreWithSelector.setState({ fur: false });
```

What I'm expecting to happen is that only the changes to state.paw are going to be logged to the console, so basically two logs. One initial and one after it's set to false.

What's actually happening is that I get three logs because I'm getting the state logged when fur changes, too.

In your docs it's a similar example:
> // Listening to selected changes, in this case when "paw" changes
> const unsub2 = useStore.subscribe(console.log, state => state.paw)

Here's a console screenshot:
<img width="326" alt="image" src="https://user-images.githubusercontent.com/15145968/197197669-94824649-e732-4e5c-8e24-f64952e6349b.png">

However, that example also fails. And what's also failing in your docs is copying the whole segment and running it in a sandbox because of the use of shallow variable and not importing create first. I get that this second point is a bit nitpicky maybe but it would be neat to copy a block of code without running into errors first.

**Comment by dai-shi** [comment]
I think that doc is under construction.

Please see README.md: https://github.com/pmndrs/zustand#using-subscribe-with-selector

You are very welcome to contribute to fix the docs.

**Reply by guenyoo** [reply]
I see, thanks a lot. 

For everyone else coming across this issue, this works fine:
```javascript
import create from "zustand";
import { subscribeWithSelector } from 'zustand/middleware'

const useDogStore = create(
  subscribeWithSelector(() => ({ paw: true, snout: true, fur: true }))
)

// Listening to selected changes, in this case when "paw" changes
const unsub2 = useDogStore.subscribe((state) => state.paw, console.log)
// Subscribe also exposes the previous value
const unsub3 = useDogStore.subscribe(
  (state) => state.paw,
  (paw, previousPaw) => console.log({paw, previousPaw})
);

useDogStore.setState({paw: false});
useDogStore.setState({snout: false});
useDogStore.setState({fur: false});
```

and logs:
<img width="305" alt="image" src="https://user-images.githubusercontent.com/15145968/197474555-2d922773-444b-4ee9-8b57-1f92aadb57b5.png">


@dai-shi I would love to help with the documentation example but I haven't done anything like that before. Is there a guide on how to contribute somewhere?

**Comment by dai-shi** [comment]
> I would love to help with the documentation example but I haven't done anything like that before. Is there a guide on how to contribute somewhere?

@guenyoo Please join this thread https://github.com/pmndrs/zustand/discussions/1033#discussioncomment-3017730 and say you want to contribute. I hope @chrisk-7777 can assist you.

**Comment by chrisk-7777** [comment]
Thanks @guenyoo ! I've @ 'ed you over on the main discussion thread about docs. 
Please don't hesitate to let me know if you have any questions

**Reply by guenyoo** [reply]
Will do! Once I‘m finished with work I‘ll try to get my bearings and will ask you how to contribute. Thanks for the quick replies!
